org.sailfishos.usermanager.xml

```diff
@@ -0,0 +1,35 @@
+<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
+"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<node>
+  <interface name="org.sailfishos.usermanager">
+    <method name="users">
+        <arg direction="out" type="a(ss)" name="users"/>
+        <annotation name="org.qtproject.QtDBus.QtTypeName.Out0" value="QList&lt;SailfishUser&gt;"/>
+    </method>
+    <method name="addUser">
+        <arg direction="in" type="s" name="login_name"/>
+        <arg direction="in" type="s" name="user_name"/>
+    </method>
+    <signal name="userAdded">
```

Comment:
This could also send id so that there is all the information about the user.

Reply:
Done

Reply:
uid is not used elsewhere in this API, and is easy enough to look up if necessary with `getpwnam`. If uid needs to be covered by this API too, then it needs to be added to `modifyUser` and `userModified` too.

Reply:
Well, true. And the other way around using getpwuid is probably not more efficient either.

Something I started wondering now is that why this doesn't send SailfishUser instance? That's anyway what users returns.

Reply:
Now sends the whole struct

Reply:
I know I'm going forwards and backwards and whatever but... would it be better interface design if it just used usernames to identify users without these structs and then the client could check the information with getpwnam..?

Reply:
Too bad that Unix password database doesn't have a clearly defined key. At any point of time both username and uid have to be unique, but either one can be changed and you can't rely on any of them as an immutable key.

but...

If we are to pick something as a key, that better be uid, I think. Do we support changing uid on the fly (including fixing file system permissions) in the first place? If not then let's use uid for identifying the users. As a bonus, it's smaller and comparing numbers is easier than comparing strings, no need for character set conversions and so on. Let's use numbers. Processors are better suited for handling numbers :)

Reply:
No, we don't support changing uid and don't wish to implement that.

I don't know how glibc works internally regarding these getpwnam and getpwuid but text files are text so it needs convert the uids in the file to integers or alternatively convert the uid to a string and compare strings anyway if getpwuid is used. So I'm not sure it helps much here, although otherwise I agree. The internal representation on UI side uses uids anyway.

Also something I thought about, I wonder if getpw{name,uid} needs to go through the whole file every time or if stores some internal state to make it faster. There needs to be a few successive calls to them anyway if one wants to read all users. It probably doesn't matter much in practice as the files are not that big but it affects the time complexity still.

Reply:
Since libuser already has all these fields in it's entity strruct they may be sent to the  client as well, very little overhead. So at the moment I'm keeping this signal as it is.